File: Dp/10. Regular Expression Matching/[Like And Useful] Regular Expression Matching.py
# hard to thought.
# always think about state machine...
# time: O(N^2). space: O(N^2)
'''
  base if p is end. check s is end or not.
  then check j char.
  if j+1 char == '*'. then can have two case: ignore, or j ok continue use.
  else just still go i+1, j+1.
'''

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        
        @lru_cache(None)
        def dfs(i, j):
            if j == len(p): return i == len(s)
            first = i < len(s) and p[j] in {s[i], '.'}
            if j+1 < len(p) and p[j+1] == '*':
                ans = dfs(i, j+2) or first and dfs(i+1, j)
            else:
                ans = first and dfs(i+1, j+1)
            return ans
        
        return dfs(0, 0)

File: Dp/10. Regular Expression Matching/[Like And Useful] Regular Expression Matching.cpp
// Top-down is very simple.
// so base case is j == pattern.size, check i == string.size
// and first check i==j or j is '.'
// then 2 case. if p[j+1] == '*'.
// we can delete then dfs(i, j+2) or we can first (i+1, j).
// 
// O(N^2)

class Solution {
public:
    int dp[35][35];
    string s, p;
    
    bool dfs(int i, int j){
        if(j == p.size()) return i == s.size();
        if(dp[i][j] != -1) return dp[i][j];
        bool res = false;
        bool first = i < s.size() && (p[j] == '.' | p[j] == s[i]);
        if(j + 1 < p.size() && p[j+1] == '*')
            res = dfs(i, j+2) | (first && dfs(i+1, j));
        else
            res = first && dfs(i+1, j+1);
        return dp[i][j] = res;
    }
    
    bool isMatch(string s, string p) {
        memset(dp, -1, sizeof dp);
        this->s = s, this->p = p;
        return dfs(0, 0);
    }
};
